http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/DbReference.scala

```diff
@@ -43,42 +43,47 @@ object DbReference extends LazyLogging {
           }
         }
         throw e
-    } finally {
-      dbConnection.close()
     }
   }
 
-  def init(config: LiquibaseConfig)(implicit executionContext: ExecutionContext): DbReference = {
+  def init[F[_]: Async: ContextShift](config: LiquibaseConfig, concurrentDbAccessPermits: Semaphore[F], blocker: Blocker): Resource[F, DbReference[F]] = {
     val dbConfig =
       DatabaseConfig.forConfig[JdbcProfile]("mysql", org.broadinstitute.dsde.workbench.leonardo.config.Config.config)
 
-    if (config.initWithLiquibase)
-      initWithLiquibase(dbConfig.db.source, config)
-
-    DbReference(dbConfig)
+    for {
+      db <- Resource.make(Async[F].delay(dbConfig.db))(db => Async[F].delay(db.close()))
+      dbConnection <- Resource.make(Async[F].delay(db.source.createConnection()))(conn => Async[F].delay(conn.close()))
+      initLiquidbase = if (config.initWithLiquibase) Async[F].delay(initWithLiquibase(dbConnection, config)) else Async[F].unit
+      _ <- Resource.liftF(initLiquidbase)
+    } yield new DbRef[F](dbConfig, db, concurrentDbAccessPermits, blocker)
   }
 }
 
-case class DbReference(private val dbConfig: DatabaseConfig[JdbcProfile])(
-  implicit val executionContext: ExecutionContext
-) {
-  val dataAccess = new DataAccess(dbConfig.profile)
-  val database: JdbcBackend#DatabaseDef = dbConfig.db
+trait DbReference[F[_]] {
+  def dataAccess: DataAccess
+  def inTransaction[T](
+                        dbio: DBIO[T],
+                        isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead
+                      ): F[T]
+}
+
+private class DbRef[F[_]: Async: ContextShift](dbConfig: DatabaseConfig[JdbcProfile], database: JdbcBackend#DatabaseDef, concurrentDbAccessPermits: Semaphore[F], blocker: Blocker) extends DbReference[F] {
+  val dataAccess = new DataAccess(dbConfig.profile, blocker)
 
-  def inTransaction[T](f: (DataAccess) => DBIO[T],
+  private def inTransactionFuture[T](dbio: DBIO[T],
                        isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead): Future[T] = {
     import dataAccess.profile.api._
-    database.run(f(dataAccess).transactionally.withTransactionIsolation(isolationLevel))
+    database.run(dbio.transactionally.withTransactionIsolation(isolationLevel))
   }
 
-  def inTransactionIO[T](
-    f: (DataAccess) => DBIO[T],
+  def inTransaction[T](
+    dbio: DBIO[T],
     isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead
-  )(implicit cs: ContextShift[IO]): IO[T] =
-    IO.fromFuture(IO(inTransaction(f, isolationLevel)))
+  ): F[T] = concurrentDbAccessPermits.withPermit(blocker.blockOn(Async.fromFuture(Async[F].delay(inTransactionFuture(dbio, isolationLevel)))))
```

Comment:
So if I'm understanding this right this runs DB operations on a Blocker backed by a `cachedThreadPool` -- but with a semaphore enforcing a max concurrent access of 1000.  This seems reasonable to me.

Reply:
yep